### PR TITLE
[FIRRTL] Add black box source helper pass

### PR DESCRIPTION
**Pull request by fabianschuiki**
This PR does the following:

* Add the `BlackBoxSourceHelper` transformation pass which is modeled after the corresponding Scala implementation. It honors a set of FIRRTL annotations that declare source code for black boxes to be copied to an output directory. This can be done inline, as a separate path, or the rather Java-specific resource mechanism.

I tried to stay faithful to the original Scala implementation. Since the pass interacts with additional files specified by the user in the input source, I added the ability to make it treat input files relative to some path (usually the directory around the input FIR file) and output files relative to some path (usually the output directory for split Verilog). This should remove a lot of confusion by making the operation invariant to the current working directory.

I'm pretty sure my handling of the additional options/arguments to the pass is not very idiomatic LLVM code. The pass specifies an `input-prefix`, `output-prefix`, and `resource-prefix` option. These are generated as part of the pass `...Base`, but I'm not sure how something like `firtool` is supposed to configure these programmatically. I'm aware of the nice `--pass-pipeline=foo{x=y}` syntax that allows the user to configure things, but what if the tool needs to set the options based on some other user input (input and output file names in this case)?

The original Scala implementation is here: https://github.com/chipsalliance/firrtl/blob/master/src/main/scala/firrtl/transforms/BlackBoxSourceHelper.scala

### Todo
- [x] Incorporate #953
- [x] Incorporate #973

### CODE REVIEW

File: include/circt/Dialect/FIRRTL/Passes.td [hunk 1]
```diff
@@ -75,4 +75,69 @@ def ExpandWhens : Pass<"firrtl-expand-whens", "firrtl::FModuleOp"> {
...
+      ```
+      Specifies the file `path` as source code for the module. Copies the file
+      to the target directory.
+    - ```
+      {
+        class = "firrtl.transforms.BlackBoxResourceAnno",
+        resourceId = "myfile.v"
+      }
+      ```
+      Specifies the file `path` as soruce code for the module. In contrast to
```
Comment (darthscsi) [inline include/circt/Dialect/FIRRTL/Passes.td:121]: source

File: test/firtool/file-list.mlir [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  // VERILOG-TOP-NEXT:  endmodule
+  firrtl.module @test_mod() {
+    firrtl.instance @ExtInline {name = "foo", portNames = []}
+    firrtl.instance @ExtResource {name = "bar", portNames = []}
+    firrtl.instance @ExtPath {name = "gib", portNames = []}
+  }
+
+  // VERILOG-FOO-LABEL: module ExtInline(); endmodule
+  // VERILOG-HDR-LABEL: `define SOME_MACRO
+  firrtl.extmodule @ExtInline() attributes {annotations = [
```
Comment (darthscsi) [inline test/firtool/file-list.mlir:36]: Firrtl really makes a single module to hold multiple blackboxs?

**Review by darthscsi — COMMENTED** [review]
First, please split the .f generation and ignoring extmodules with split verilog into a separate patch.

Second, this needs splitting so input and output are not coupled.  I would have a pass that consumes the annotations and injests the files into a verbatim (or a new sv.verbatim.file or sv.module.verbatim) op.  Then when the ir is saved to disk, the things to copy to the output are saved too.  Export Verilog then can operate on just the IR and still produce all the blackboxes.

Even if files are copied into the ir, we should produce a resolved set of files in the IR in nodes that export verilog can deal with.

And by export verilog, I mean a pass running in the verilog emission phase of firtool, it's completely reasonable that dealing with the output of these black boxes is it's own pass.

File: test/firtool/file-list.mlir (hunk 2 above)
Comment (fabianschuiki) [reply test/firtool/file-list.mlir:36]: The blackbox source code is annotated by the programmer in Chisel, and they're free to add multiple such annotations. There's some special casing of `*.svh` even that suggests that multiple annotations does occur in practice.

**Comment by fabianschuiki** [comment]
> First, please split the .f generation and ignoring extmodules with split verilog into a separate patch.

Done, see #953.

**Comment by fabianschuiki** [comment]
> Second, this needs splitting so input and output are not coupled. I would have a pass that consumes the annotations and injests the files into a verbatim (or a new sv.verbatim.file or sv.module.verbatim) op. Then when the ir is saved to disk, the things to copy to the output are saved too. Export Verilog then can operate on just the IR and still produce all the blackboxes.
> 
> Even if files are copied into the ir, we should produce a resolved set of files in the IR in nodes that export verilog can deal with.

That's a nice way of dealing with it. I could definitely see some concerns regarding large black box files come up in the future, but we can deal with that once it becomes a problem. I applied your suggestions: added a `sv.verbatim.file` operation, renamed the pass to `BlackBoxReader`, and modified it to generate these inline file operations. Verilog export the writes these to disk (went well with the existing Verilog emission with only few necessary changes).

File: test/firtool/file-list.mlir (hunk 2 above)
Comment (darthscsi) [reply test/firtool/file-list.mlir]: It seems hacky to allow multiple files be generated for a single extmodule with a big attribute list.  When we lower to RTL, we should fix that, but if multi-file attributes exist in firrtl, we might as well preserver that in the firrtl dialect.

Comment (fabianschuiki) [reply test/firtool/file-list.mlir]: I agree that it's not very elegant. But I could totally see it being necessary for some cases, e.g., if you have a IP-vendor-provided black box source file with a ``` `include "foo.svh" ```, forcing you to provide some config in `foo.svh` as well. 

@seldridge: Is this feature used in Chisel sources and should we preserve it?

Actually on second thought: Disallowing multiple file annotations would require adding more code to explicitly prevent a technically legitimate use case (and does not simplify downstream processing as far as I can see). Wouldn't that be a somewhat arbitrary restriction on the FIRRTL dialect with no technical merit?

Comment (seldridge) [reply test/firtool/file-list.mlir]: Yes, this should support multiple resources. E.g.: https://github.com/chipsalliance/rocket-chip/blob/147bdcc4a26c74e5d7a47e3d667d456699d6d11f/src/main/scala/devices/debug/Periphery.scala#L204. 

This is more a kludge of "add this file to the output directory" than it is about specifying the implementation of the blackbox. The implementation may be one file or multiple files. This borders on build system stuff.

**Comment by fabianschuiki** [comment]
Opened #973 which introduces a dedicated `sv.file` operation. Once that lands this PR can use a `sv.file { sv.verbatim "..." }` combo instead of introducing a `sv.verbatim.file` operation. That also handles file emission in `ExportVerilog` properly without the need for further changes. Reduces clutter overall.

File: test/firtool/file-list.mlir (hunk 2 above)
Comment (fabianschuiki) [reply test/firtool/file-list.mlir]: Since we have use cases for this in Chisel, it's probably easier to just support it and then make a conscious effort to phase things like simulation source file inclusion out upstream. Otherwise users get deterred by warnings like "you use multiple black box source files here".

File: tools/firtool/firtool.cpp [hunk 3]
```diff
@@ -91,6 +91,12 @@ static cl::opt<bool>
                        cl::desc("ignore the @info locations in the .fir file"),
                        cl::init(false));
 
+static cl::opt<std::string> blackBoxResourcePath(
+    "blackbox-resource-path",
```
Comment (lattner) [inline tools/firtool/firtool.cpp:95]: I think this can be declares as a pass option, which would be more self contained than this.

File: include/circt/Dialect/SV/SVStatements.td [hunk 4]
```diff
@@ -437,6 +441,20 @@ def VerbatimOp : SVOp<"verbatim"> {
   ];
 }
 
+def VerbatimFileOp : SVOp<"verbatim.file"> {
```
Comment (lattner) [inline include/circt/Dialect/SV/SVStatements.td:444]: Please split IR additions out to their own patch so we get high visibility on them.

File: include/circt/Dialect/SV/SVStatements.td [hunk 5]
```diff
@@ -437,6 +441,20 @@ def VerbatimOp : SVOp<"verbatim"> {
...
+  let summary = "Verbatim opaque text emitted into a separate file.";
+  let description = [{
+    This operation produces opaque text in a separate SystemVerilog output file.
+    The `fileName` is a file path relative to other output where the `string`
+    will be written to.
+
+    Emitters honor a `nofilelist` unit attribute which if present prevent the
```
Comment (lattner) [inline include/circt/Dialect/SV/SVStatements.td:451]: Shouldn't `nofilelist` attribute be declared in ODS?

File: lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+  BlackBoxReaderPass(llvm::Optional<StringRef> inputPrefix = {},
+                     llvm::Optional<StringRef> resourcePrefix = {}) {
+    if (inputPrefix.hasValue())
+      this->inputPrefix = inputPrefix.getValue().str();
+    if (resourcePrefix.hasValue())
+      this->resourcePrefix = resourcePrefix.getValue().str();
+  }
+
+  /// Emit the annotated source code for black boxes in a circuit.
+  void runOnOperation() override {
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp:52]: Please move the bodies of these large methods out of line,  this makes it a bit easier to read the class declaration and reduces nesting of their bodies.

File: lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+  } else {
+    llvm::sys::path::append(base, suffix);
+  }
+}
+
+namespace {
+
+struct BlackBoxReaderPass : public BlackBoxReaderBase<BlackBoxReaderPass> {
+  BlackBoxReaderPass(llvm::Optional<StringRef> inputPrefix = {},
+                     llvm::Optional<StringRef> resourcePrefix = {}) {
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp:44]: Would it be reasonable to eliminate all the Optionals, and just pass down the missing case as an empty string? (e.g. all the way from firtool).

This simplifies the code and it looks like you're throwing away the none/empty difference anyway.

File: lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// Read Verilog source files for black boxes based on corresponding black box
+// annotations on the circuit and modules. Primarily based on:
+//
+// https://github.com/chipsalliance/firrtl/blob/master/src/main/scala/firrtl/transforms/BlackBoxSourceHelper.scala
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp:11]: nit: wrap to 80 columns

File: lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+    resourceFileNameAnnoClass =
+        StringAttr::get(cx, "firrtl.transforms.BlackBoxResourceFileNameAnno");
+    inlineAnnoClass =
+        StringAttr::get(cx, "firrtl.transforms.BlackBoxInlineAnno");
+    pathAnnoClass = StringAttr::get(cx, "firrtl.transforms.BlackBoxPathAnno");
+    resourceAnnoClass =
+        StringAttr::get(cx, "firrtl.transforms.BlackBoxResourceAnno");
+
+    // Run on the CircuitOp in the module.
+    for (auto &op : *getOperation().getBody())
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp:66]: There will be only one, but please declare the pass as running on CircuitOp instead of ModuleOp.

File: lib/Translation/ExportVerilog/ExportVerilog.cpp [hunk 10]
```diff
@@ -2923,53 +2930,71 @@ void SplitEmitter::emitModule(const LoweringOptions &options,
   auto op = mod.op;
 
   // Given the operation, determine the file stem name and how to emit it.
-  std::function<void(VerilogEmitterState &)> emit;
+  std::function<void(VerilogEmitterState &)> emitWithinFileOps;
+  std::function<void(llvm::raw_ostream & os)> emit =
+      [&](llvm::raw_ostream &os) {
```
Comment (lattner) [inline lib/Translation/ExportVerilog/ExportVerilog.cpp:2935]: When you split the new IR construct out to its own patch, please add a self contained testcase for it.

File: include/circt/Dialect/SV/SVStatements.td [hunk 11]
```diff
@@ -437,6 +441,20 @@ def VerbatimOp : SVOp<"verbatim"> {
...
+    file from showing up in generated file lists.
+  }];
+  let arguments = (ins StrAttr:$fileName, StrAttr:$string);
```
Comment (lattner) [inline include/circt/Dialect/SV/SVStatements.td:454]: I don't really understand what is being modeled, here, but we can discuss this when it gets split out.

File: tools/firtool/firtool.cpp (hunk 3 above)
Comment (fabianschuiki) [reply tools/firtool/firtool.cpp:95]: It already is, but I can remove this option on firtool altogether, since it's more on the esoteric side. If someone then wants to resolve the weird Java class resource paths, they can just add an option to the pass (that requires specifying a custom pass pipeline, right?).

File: include/circt/Dialect/SV/SVStatements.td (hunk 4 above)
Comment (fabianschuiki) [reply include/circt/Dialect/SV/SVStatements.td:444]: I have forked this as a more orthogonal `sv.file` into #973 as discussed above. Forgot to mark this PR as draft afterwards, as this PR rides on the addition of `sv.file` to emit its auxiliary files.

**Comment by fabianschuiki** [comment]
Updated this PR to leverage the new `OutputFileAttr` we gained from #973. That makes this PR a bit cleaner and more focused on the core task at hand.

File: include/circt/Dialect/FIRRTL/Passes.td [hunk 12]
```diff
@@ -111,4 +111,67 @@ def InferWidths : Pass<"firrtl-infer-widths", "firrtl::CircuitOp"> {
...
+      search path. This is a remnant of the Scala origins of FIRRTL. Copies the
+      file to the target directory.
+  }];
+
+  let constructor = "circt::firrtl::createBlackBoxReaderPass()";
+  let options = [
+    Option<"inputPrefix", "input-prefix", "std::string", "",
+      "Prefix for input paths in black box annotations. This should be the"
+      "directory where the input file was located, to allow for annotations"
+      "relative to the input file.">,
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/Passes.td:168]: Nit: I think this needs spaces to render properly:

```suggestion
      "Prefix for input paths in black box annotations. This should be the "
      "directory where the input file was located, to allow for annotations "
      "relative to the input file.">,
```

File: include/circt/Dialect/FIRRTL/Passes.td [hunk 13]
```diff
@@ -111,4 +111,67 @@ def InferWidths : Pass<"firrtl-infer-widths", "firrtl::CircuitOp"> {
...
+    Option<"resourcePrefix", "resource-prefix", "std::string",
+      "\"src/main/resources\"",
+      "Search path for black box sources specified via the"
+      "`BlackBoxResourceAnno` annotation.">
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/Passes.td:172]:
```suggestion
      "Search path for black box sources specified via the "
      "`BlackBoxResourceAnno` annotation.">
```

File: lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+  // Internalize some string attributes for easy reference later.
+  auto cx = &getContext();
+  targetDirAnnoClass =
+      StringAttr::get(cx, "firrtl.transforms.BlackBoxTargetDirAnno");
+  resourceFileNameAnnoClass =
+      StringAttr::get(cx, "firrtl.transforms.BlackBoxResourceFileNameAnno");
+  inlineAnnoClass = StringAttr::get(cx, "firrtl.transforms.BlackBoxInlineAnno");
+  pathAnnoClass = StringAttr::get(cx, "firrtl.transforms.BlackBoxPathAnno");
+  resourceAnnoClass =
+      StringAttr::get(cx, "firrtl.transforms.BlackBoxResourceAnno");
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp:101]: I hadn't thought to do it this way. Normally, I'll compare the "class" key `StringAttr` _value_ against a `char *`. I assume that by constructing these `StringAttr`'s directly, you can do the comparison by pointer comparison?

Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp:101]: There is also a benefit of you can then do the `cls == anno.getAs<A>("...")` below which then doesn't require a check before a `getValue` to avoid the null dereference. I'm good with this the way it is. :+1:

File: lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+
+      // Handle target dir annotation.
+      if (cls == targetDirAnnoClass) {
+        auto x = anno.getAs<StringAttr>("targetDir");
+        if (!x) {
+          circuitOp->emitError(targetDirAnnoClass.getValue() +
+                               " annotation missing \"targetDir\" attribute");
+          signalPassFailure();
+          continue;
+        }
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp:123]: Thanks for being pedantic and checking this stuff and throwing good errors if mandatory fields are missing.

File: test/firtool/blackbox.mlir [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+// LIST-BLACK-BOX:      magic/blackbox-inline.v
+// LIST-BLACK-BOX-NEXT: magic/blackbox-resource.v
+// LIST-BLACK-BOX-NEXT: magic/blackbox-path.v
+
+firrtl.circuit "test_mod" attributes {annotations = [
+  // Black box processing should honor only the last annotation.
+  {class = "firrtl.transforms.BlackBoxTargetDirAnno", targetDir = "ignore_me_plz"},
+  {class = "firrtl.transforms.BlackBoxTargetDirAnno", targetDir = "magic"},
+  {class = "firrtl.transforms.BlackBoxResourceFileNameAnno", resourceFileName = "definitely_bad.f"},
+  {class = "firrtl.transforms.BlackBoxResourceFileNameAnno", resourceFileName = "magic.f"}
```
Comment (seldridge) [inline test/firtool/blackbox.mlir:21]: Ugh... This is the right behavior by my read of the Scala, but I hate this. It means that annotation order becomes part of the API. 😬 😬 😬 Good catch to check this.

File: lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+  if (!emittedFiles.empty()) {
+    auto trueAttr = BoolAttr::get(cx, true);
+    std::string output;
+    llvm::raw_string_ostream os(output);
+    for (auto &file : emittedFiles) {
+      const auto &fileName = file.fileName.getValue();
+      // Exclude Verilog header files since we expect them to be included
+      // explicitly by compiler directives in other source files.
+      auto ext = llvm::sys::path::extension(fileName);
+      bool exclude = (ext == ".h" || ext == ".vh" || ext == ".svh");
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp:175]: I didn't realize this was something the Blackbox transform would do. Nice work staying true to the Scala code. :+1:

**Review by seldridge — APPROVED** [review]
This looks excellent!

I'm fine for you to land this whenever you're ready.

File: lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp (hunk 14 above)
Comment (fabianschuiki) [reply lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp:101]: Yeah exactly, this way it's just a pointer comparison. It's probably a bit esoteric to do this, but it simplifies the check further down as you say.

File: lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp (hunk 15 above)
Comment (fabianschuiki) [reply lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp:123]: :+1: I was thinking about maybe using the TableGen mechanism to generate `StructAttr` to do the parsing of the annotation dictionaries, but that just asserts if data is missing, rather than reporting back to the user. So I'm sticking with this at the moment, but it might be worthwhile contributing some code upstream that produces fallible casts for struct attrs that can properly report to the user if a dictionary should have a certain shape, but doesn't.

**Review by darthscsi — APPROVED** [review]

File: test/firtool/blackbox.mlir (hunk 16 above)
Comment (fabianschuiki) [reply test/firtool/blackbox.mlir:21]: Yeah that `fold` in the Scala code is magic :mage:

File: lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+      return signalPassFailure();
+    }
+
+    // Gather the relevant annotations on all modules in the circuit so we can
+    // process them in parallel afterwards.
+    OpBuilder builder(&getContext());
+    builder.setInsertionPointToEnd(getOperation().getBody());
+
+    for (auto &op : *circuitOp.getBody()) {
+      if (!isa<FModuleOp>(op) && !isa<FExtModuleOp>(op))
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp]: The Scala implementation says ["Handle source for Verilog ExtModules (BlackBoxes)"](https://github.com/chipsalliance/firrtl/blob/master/src/main/scala/firrtl/transforms/BlackBoxSourceHelper.scala#L52), but the actual implementation just gathers the annotations which can [target anything with a `ModuleName`](https://github.com/chipsalliance/firrtl/blob/master/src/main/scala/firrtl/transforms/BlackBoxSourceHelper.scala#L20). I figure in practice people will only put this on `ExtModule`s, but this prevents any surprises where SFC accepts something which MFC rejects. Especially since we actually don't really rely on the operation the annotation sits on.

File: lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp (hunk 15 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp:123]: Agree. This pattern shows up everywhere, e.g., [in Grand Central annotation parsing](https://github.com/llvm/circt/blob/main/lib/Dialect/FIRRTL/Import/FIRAnnotations.cpp#L99). Asserts won't cut it as you want a real error message.

The other problem here is that the location really should be in the input JSON. However, there's no support that I'm aware of for getting source locators out of `llvm::json`. However, it would eventually be cool if we can include source locators pointing at JSON that can be used to generate top-tier error messages.

File: lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp (hunk 14 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp:101]: Nice solution. I should steal this pattern for the grand central annotation handling. 😄

File: lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp (hunk 15 above)
Comment (fabianschuiki) [reply lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp:123]: Yeah that would be great to have!
